### PR TITLE
Prp 224 - FileUpload component

### PR DESCRIPTION
**Pull request by microwavenby**
## Ticket

https://wicmtdp.atlassian.net/browse/PRP-224

## Changes
* FileUploader component
* FilePreview component
* FileInput component
* Upload page draft
* Changes to add FileUploader to storybook
* Unit tests
* Playwright tests to test behavior and validation of file type / size / count

## Testing

demo-file-component

### CODE REVIEW

**Review by makaelastephens — COMMENTED** [review]
Can we make the labels default size, rather than large? 
image

File: participant/Dockerfile.playwright [hunk 1]
```diff
@@ -2,8 +2,11 @@ FROM mcr.microsoft.com/playwright:v1.31.0-focal
 # npm install must be run in a non-root directory
 WORKDIR /srv
 # Install packages needed to run playwright tests
-RUN npm install @playwright/test @axe-core/playwright uuid
+RUN npm install @playwright/test @axe-core/playwright uuid js-image-generator tmp
 # Copy playwright config and tests
 COPY playwright.config.ts .
+COPY e2e e2e
```
Comment (rocketnova) [inline participant/Dockerfile.playwright]: If you're going to `COPY` the tests in, then you should probably pull the bind mount out of the docker-compose.e2e.yml file to help reduce future confusion.

File: participant/public/locales/en/common.json [hunk 2]
```diff
@@ -96,6 +96,20 @@
     "legend": "What's your name?",
     "preferred": "Preferred name (optional)"
   },
+  "FileUploader": {
```
Comment (rocketnova) [inline participant/public/locales/en/common.json:99]: @makaelastephens @danielladevera-nava Could you proof this copy? It's a little consistent around casing (e.g. "Remove File" vs "Upload files") and periods (e.g. "This is not a valid file type." vs "This file is over the maximum size").

File: participant/public/locales/en/common.json [hunk 3]
```diff
@@ -96,6 +96,20 @@
...
+    "selected": "Selected file",
+    "label": "Upload files",
+    "document": "Document",
+    "ifNeeded": "(if needed)",
+    "noFiles": "Choose or drag a file to upload",
+    "additionalFiles": "Choose or drag another file to upload",
+    "altText": "Image preview for",
+    "fileTypeError": "This is not a valid file type.",
+    "fileSizeError": "This file is over the maximum size",
+    "fileCountError": "You have reached the maximum number of files; <br/> please remove files before adding additional files",
```
Comment (rocketnova) [inline participant/public/locales/en/common.json]: @makaelastephens @danielladevera-nava Can this be a single line? 

@microwavenby If it needs to be 2 lines, can we use a `<span>` with `display: block` css class instead? We should use css to control visual display

File: participant/styles/_uswds-theme-custom-styles.scss [hunk 4]
```diff
@@ -41,11 +41,28 @@ i.e.
   flex: 1;
 }
 
+
 // Custom classes
 p {
   @include typeset('body');
 }
 
+// File Input styling
```
Comment (rocketnova) [inline participant/styles/_uswds-theme-custom-styles.scss]: Nit: Would be more consistent with the rest of the file to put this section after the custom css classes with a comment like this:

```suggestion
// USWDS File Input:
// - Fix issue where CSS captures mouse clicks
// - Change styling
```

File: participant/styles/_uswds-theme-custom-styles.scss [hunk 5]
```diff
@@ -41,11 +41,28 @@ i.e.
...
+.usa-file-input__preview-heading {
+  pointer-events: auto;
+}
+
+.usa-file-input__target:hover {
+  border-color: #1a4480;
+}
+
+.file-input-error {
+  font-weight: bold;
```
Comment (rocketnova) [inline participant/styles/_uswds-theme-custom-styles.scss]: Nit:
```suggestion
  @include u-text('bold');
```

File: participant/app/routes/$localAgency/recertify/upload.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    // No empty files from the real component sneaking in
+    data.delete("documents");
+    fileInputRef.current?.files.forEach((value) => {
+      data.append("documents", value);
+    });
+    formSubmit(data, { method: "post", encType: "multipart/form-data" });
+  };
+  return (
+    <div>
+      <h1>{t("Upload.title")}</h1>
```
Comment (rocketnova) [inline participant/app/routes/$localAgency/recertify/upload.tsx:64]:
```suggestion
      <h1><Trans i18nKey="Upload.title" /></h1>
```

File: participant/app/routes/$localAgency/recertify/upload.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      <Form
+        method="post"
+        id="uploadForm"
+        encType="multipart/form-data"
+        className="usa-form usa-form--large"
+        name="documents-form"
+        onSubmit={(event) => handleSubmit(event)}
+      >
+        <FileUploader {...defaultProps} ref={fileInputRef} />
+        <Button type="submit" value="action" name="action">
```
Comment (rocketnova) [inline participant/app/routes/$localAgency/recertify/upload.tsx]: What is the relationship between the `onSubmit()` and the `Button` action? I'm having a little bit of a hard time piecing it together.

File: participant/app/components/internal/FilePreview.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
+/* Lifted and edited from @trussworks/react-uswds FilePreview internal only component
+ * This component (in the source) gets displayed as part of the @trussworks/react-uswds
+ * FileInput component after a file (or files) are selected
```
Comment (rocketnova) [inline participant/app/components/internal/FilePreview.tsx:3]: What have we edited?

File: participant/app/components/FileUploader.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  const [files, setFiles] = useState<FileState>({});
+  const [previews, setPreviews] = useState(<></>);
+  const fileInputRef = useRef(null);
+  const documentString = t(`${labelKey}.document`);
+  const [errorMessage, setErrorMessage] = useState("");
+
+  const ifNeeded = t(`${labelKey}.ifNeeded`);
+  const currentDocumentNumber = Object.keys(files).length;
+
+  useImperativeHandle(
```
Comment (rocketnova) [inline participant/app/components/FileUploader.tsx:70]: What does this do?

File: participant/app/components/FileUploader.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  };
+  // eslint-disable-next-line   react-hooks/exhaustive-deps -- (deps list is correct)
+  useEffect(renderPreviews, [files, documentString, labelKey]);
+  const fileTypeCheck = (newFile: File): boolean => {
+    if (!accept || accept == "*") {
+      return true;
+    }
+    const acceptedTypes = accept.split(",");
+    let fileTypeAllowed = true;
+    if (fileTypeAllowed) {
```
Comment (rocketnova) [inline participant/app/components/FileUploader.tsx]: Is this check needed? Won't this always be `true`?

File: participant/app/components/FileUploader.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+      for (let j = 0; j < acceptedTypes.length; j += 1) {
+        const fileType = acceptedTypes[parseInt(`${j}`)];
+        fileTypeAllowed =
+          newFile.name.indexOf(fileType) > 0 ||
+          newFile.type.includes(fileType.replace(/\*/g, ""));
```
Comment (rocketnova) [inline participant/app/components/FileUploader.tsx]: Can you walk me through what this does? What is the expected structure of `newFile`?

**Comment by rocketnova** [comment]
@microwavenby Mostly nits and clarifying questions.

File: participant/Dockerfile.playwright (hunk 1 above)
Comment (microwavenby) [reply participant/Dockerfile.playwright]: I s2g i don't even remember adding this

File: participant/public/locales/en/common.json (hunk 3 above)
Comment (microwavenby) [reply participant/public/locales/en/common.json]: I'll do whatever y'all want.

File: participant/app/routes/$localAgency/recertify/upload.tsx (hunk 7 above)
Comment (microwavenby) [reply participant/app/routes/$localAgency/recertify/upload.tsx]: ... This is a habit from the eligibility screener - to determine the difference in the Action between a new submission and an update (like in the review flow)

It probably doesn't make sense to have the value be "action", it should be "submit"
But, the button's value gets submitted with the rest of the form
Like:
`action=submit` (if i change the `value="submit"`)

File: participant/app/components/FileUploader.tsx (hunk 11 above)
Comment (microwavenby) [reply participant/app/components/FileUploader.tsx]: The structure? Or the type? It's a `File` object

File: participant/app/components/FileUploader.tsx (hunk 10 above)
Comment (microwavenby) [reply participant/app/components/FileUploader.tsx]: yep that's some 🍝

File: participant/public/locales/en/common.json (hunk 3 above)
Comment (microwavenby) [reply participant/public/locales/en/common.json]: This is what wrapping that in a `<span className='display-block'>` does:

image

File: participant/public/locales/en/common.json (hunk 2 above)
Comment (microwavenby) [reply participant/public/locales/en/common.json:99]: I was working with copy from the Figma; there is no copy for the file component in the Lucidchart

**Review by rocketnova — APPROVED** [review]
LGTM! 🚢

**Pull request merged**